File: exercises/matmul/matmul.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define dt(start, end)  ((end.tv_sec - start.tv_sec) + \
			 1/1000000.0*(end.tv_usec - start.tv_usec))

/*
  define rows and columns of matrices a, b and c
 a[nra,nca], b[nca,ncb], c[nra, ncb]
*/

#define nra 1500
#define nca 2000
#define ncb 1000

double a[nra][nca], b[nca][ncb], c[nra][ncb]; 

int main() {
  int i,j,k;
  struct timeval icalc, scalc, ecalc;
  double flops, sum, timing ;
  
  flops = 2.0 * nra * nca * ncb;

  gettimeofday(&icalc, NULL);
  for (i = 0; i < nra; i++){
    for (j = 0; j < nca; j++){
      a[i][j] = (double)(i+j);
    }
  }
  for (j = 0; j < nca; j++){
    for (k = 0; k < ncb; k++){
      b[j][k] = (double)(i*j);
    }
  }
  for (i = 0; i < nra; i++){
    for (k = 0; k < ncb; k++){
      c[i][k] = 0.0;
    }
  }
  
  gettimeofday(&scalc, NULL);
  /* Parallelize the following block of code */
  for (i = 0; i < nra; i++){
    for (k = 0; k < ncb; k++){
      sum = 0.0;
      for (j = 0; j < nca; j++){
	sum = sum + a[i][j] * b[j][k];
      }
      c[i][k] = sum;
    }
  }
  gettimeofday(&ecalc, NULL);

  timing = dt(scalc,ecalc);
  printf("Init Time: %6.3f Calc Time: %6.3f GFlops: %7.3f\n",dt(icalc,scalc),timing,1e-9*flops/timing );
}

File: exercises/matmul/matmul.f90
program matrix_mul

  implicit none

  integer, parameter :: dp = selected_real_kind(14)
  integer :: i,j,k
  integer, parameter :: nra=1500, nca=2000, ncb=1000
  real(dp) :: a(nra,nca) , b(nca,ncb) , c(nra,ncb)
  real(dp) :: flops, sum
  real(dp) :: init_time, start_time, end_time


  flops = 2d0 * float(nra) * float(nca) * float(ncb)

  call cpu_time(init_time)
  c = 0d0

  do i = 1,nra
     do j = 1,nca
        a(i,j) = i + j
     end do
  end do
  
  do i = 1,nca
     do j = 1,ncb
        b(i,j) = i * j
     end do
  end do
  
  call cpu_time(start_time)
  ! Parallelize the following block of code
  do j = 1, nca
     do k = 1, ncb
        sum = 0d0
        do i = 1, nra
           sum = sum + a(i,j) * b(j,k) 
        end do
        c(i,k) = sum
     end do
  end do
  call cpu_time(end_time)
  
  print '(a,f6.3,a,f6.3,a,f7.3)', 'Init Time: ', start_time - init_time, &
       ' Calc Time: ', end_time - start_time, &
       ' GFlops: ', 1d-9 * flops/(end_time - start_time)
  
  
end program matrix_mul
